### PR TITLE
Add one probe integration

### PR DESCRIPTION
**Pull request by lynxx131**
Adds support for The One Probe to get recipe progress and machine energy information.

This integration is also provided for Fabric using the unofficial port.

### CODE REVIEW

**Comment by lynxx131** [comment]
PR job seems to be failing to locate a dependency:

```
Could not find com.github.LlamaLad7:MixinExtras:0.0.12
```

This also fails for me locally on the dev-1.18 branch if I remove my gradle cache so I think the MixinExtras 0.0.12 artifact is currently unavailable.

**Comment by lynxx131** [comment]
I've been able to test this against Forge but no luck with Fabric at the moment as GTI won't launch due to the following error:

```
Caused by: java.lang.NoSuchMethodError: 'net.minecraft.world.level.block.state.BlockState com.github.gregtechintergalactical.gtrubber.tree.BlockRubberLog.method_9564()'
	at com.github.gregtechintergalactical.gtrubber.tree.BlockRubberLog.<init>(BlockRubberLog.java:43)
	at com.github.gregtechintergalactical.gtrubber.GTRubberData.<clinit>(GTRubberData.java:21)
	at com.github.gregtechintergalactical.gtrubber.GTRubber.onRegistrationEvent(GTRubber.java:100)
	at muramasa.antimatter.AntimatterAPI.lambda$onRegistration$27(AntimatterAPI.java:409)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at muramasa.antimatter.AntimatterAPI.onRegistration(AntimatterAPI.java:409)
	at muramasa.antimatter.registration.fabric.AntimatterRegistration.onRegister(AntimatterRegistration.java:56)
	at muramasa.antimatter.registration.fabric.AntimatterRegistration.onRegister(AntimatterRegistration.java:47)
	at muramasa.antimatter.fabric.AntimatterImpl.initialize(AntimatterImpl.java:63)
	at muramasa.antimatter.integration.kubejs.fabric.AntimatterKubeJSImpl.onRegister(AntimatterKubeJSImpl.java:14)
	at muramasa.antimatter.integration.kubejs.AntimatterKubeJS.onRegister(AntimatterKubeJS.java)
	at muramasa.antimatter.integration.kubejs.AntimatterKubeJS.initStartup(AntimatterKubeJS.java:19)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at dev.latvian.mods.kubejs.util.KubeJSPlugins.forEachPlugin(KubeJSPlugins.java:83)
	at dev.latvian.mods.kubejs.KubeJS.<init>(KubeJS.java:117)
	at dev.latvian.mods.kubejs.fabric.KubeJSFabric.onInitialize(KubeJSFabric.java:12)
```

**Comment by lynxx131** [comment]
GTRubber dependency issue has been sorted, this PR is now ready for review.

File: fabric/src/main/java/muramasa/antimatter/integration/top/fabric/EnergyInfoProvider.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void addProbeInfo(ProbeMode mode, IProbeInfo probeInfo, Player player, Level level,
+                             BlockState blockState, IProbeHitData data) {
+
+        if (blockState.hasBlockEntity()) {
+            BlockEntity tile = Utils.getTile(level, data.getPos());
+
+            if (tile == null) return;
```
Comment (Trinsdar) [inline fabric/src/main/java/muramasa/antimatter/integration/top/fabric/EnergyInfoProvider.java]: This line is completely unnecessary, the instance check will return false if the tile is null.

File: fabric/src/main/java/muramasa/antimatter/integration/top/fabric/RecipeInfoProvider.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public void addProbeInfo(ProbeMode mode, IProbeInfo probeInfo, Player player, Level level,
+                             BlockState blockState, IProbeHitData data) {
+
+        if (blockState.hasBlockEntity()) {
+            BlockEntity tile = Utils.getTile(level, data.getPos());
+
+            if (tile == null) return;
```
Comment (Trinsdar) [inline fabric/src/main/java/muramasa/antimatter/integration/top/fabric/RecipeInfoProvider.java]: Same as abive

File: forge/src/main/java/muramasa/antimatter/integration/top/forge/EnergyInfoProvider.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void addProbeInfo(ProbeMode mode, IProbeInfo probeInfo, Player player, Level level,
+                             BlockState blockState, IProbeHitData data) {
+
+        if (blockState.hasBlockEntity()) {
+            BlockEntity tile = Utils.getTile(level, data.getPos());
+
+            if (tile == null) return;
```
Comment (Trinsdar) [inline forge/src/main/java/muramasa/antimatter/integration/top/forge/EnergyInfoProvider.java]: Same as above

File: forge/src/main/java/muramasa/antimatter/integration/top/forge/RecipeInfoProvider.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Override
+    public void addProbeInfo(ProbeMode mode, IProbeInfo probeInfo, Player player, Level level,
+                             BlockState blockState, IProbeHitData data) {
+
+        if (blockState.hasBlockEntity()) {
+            BlockEntity tile = Utils.getTile(level, data.getPos());
+
+            if (tile == null) return;
+
```
Comment (Trinsdar) [inline forge/src/main/java/muramasa/antimatter/integration/top/forge/RecipeInfoProvider.java]: Same as above

**Review by Trinsdar — CHANGES_REQUESTED** [review]

File: forge/src/main/java/muramasa/antimatter/integration/top/forge/RecipeInfoProvider.java (hunk 4 above)
Comment (lynxx131) [reply forge/src/main/java/muramasa/antimatter/integration/top/forge/RecipeInfoProvider.java]: Removed

File: forge/src/main/java/muramasa/antimatter/integration/top/forge/EnergyInfoProvider.java (hunk 3 above)
Comment (lynxx131) [reply forge/src/main/java/muramasa/antimatter/integration/top/forge/EnergyInfoProvider.java]: Removed

File: fabric/src/main/java/muramasa/antimatter/integration/top/fabric/RecipeInfoProvider.java (hunk 2 above)
Comment (lynxx131) [reply fabric/src/main/java/muramasa/antimatter/integration/top/fabric/RecipeInfoProvider.java]: Removed

File: fabric/src/main/java/muramasa/antimatter/integration/top/fabric/EnergyInfoProvider.java (hunk 1 above)
Comment (lynxx131) [reply fabric/src/main/java/muramasa/antimatter/integration/top/fabric/EnergyInfoProvider.java]: Removed

**Comment by lynxx131** [comment]
Just tried this PR in a non-dev client against the current point release of TOP and the formatting's a bit off so going to look at that.

**Review by Trinsdar — APPROVED** [review]

**Comment by lynxx131** [comment]
I've pulled Fabric integration from this PR since I can't figure out how to properly integrate the TOP fabric port.

The forge integration works and provides energy storage info, recipe progress and informs users whether a multiblock structure is valid or not.

**Comment by Trinsdar** [comment]
fyi I will probably figure out the fabric side another time, probably same time I do wthit

**Pull request merged**